src/execution/join_hashtable.cpp

```diff
@@ -794,4 +888,221 @@ idx_t JoinHashTable::FillWithHTOffsets(data_ptr_t *key_locations, JoinHTScanStat
 	}
 	return key_count;
 }
+
+void JoinHashTable::SwizzleBlocks() {
+	RowDataCollectionScanner::AlignHeapBlocks(*swizzled_block_collection, *swizzled_string_heap, *block_collection,
+	                                          *string_heap, layout);
+}
+
+void JoinHashTable::UnswizzleBlocks() {
+	auto &blocks = swizzled_block_collection->blocks;
+	auto &heap_blocks = swizzled_string_heap->blocks;
+#ifdef DEBUG
+	if (!layout.AllConstant()) {
+		D_ASSERT(blocks.size() == heap_blocks.size());
+		D_ASSERT(swizzled_block_collection->count == swizzled_string_heap->count);
+	}
+#endif
+
+	for (idx_t block_idx = 0; block_idx < blocks.size(); block_idx++) {
+		auto &data_block = blocks[block_idx];
+
+		if (!layout.AllConstant()) {
+			auto block_handle = buffer_manager.Pin(data_block->block);
+
+			auto &heap_block = heap_blocks[block_idx];
+			D_ASSERT(data_block->count == heap_block->count);
+			auto heap_handle = buffer_manager.Pin(heap_block->block);
+
+			// Unswizzle and move
+			RowOperations::UnswizzlePointers(layout, block_handle.Ptr(), heap_handle.Ptr(), data_block->count);
+			string_heap->blocks.push_back(move(heap_block));
+			string_heap->pinned_blocks.push_back(move(heap_handle));
+		}
+
+		// Fixed size stuff can just be moved
+		block_collection->blocks.push_back(move(data_block));
+	}
+
+	// Update counts and clean up
+	block_collection->count = swizzled_block_collection->count;
+	string_heap->count = swizzled_string_heap->count;
+	swizzled_block_collection->Clear();
+	swizzled_string_heap->Clear();
+
+	D_ASSERT(SwizzledCount() == 0);
+}
+
+void JoinHashTable::ComputePartitionSizes(ClientConfig &config, vector<unique_ptr<JoinHashTable>> &local_hts,
+                                          idx_t max_ht_size) {
+	external = true;
+
+	// First set the number of tuples in the HT per partitioned round
+	total_count = 0;
+	idx_t total_size = 0;
+	for (auto &ht : local_hts) {
+		// TODO: SizeInBytes / SwizzledSize overestimates size by a lot because we make extra references of heap blocks
+		//  Need to compute this more accurately
+		total_count += ht->Count() + ht->SwizzledCount();
+		total_size += ht->SizeInBytes() + ht->SwizzledSize();
+	}
+
+	if (total_count == 0) {
+		return;
+	}
+
+	total_size += PointerTableCapacity(total_count) * sizeof(data_ptr_t);
+	idx_t avg_tuple_size = total_size / total_count;
```

Comment:
Can this trigger a divide by zero?

Reply:
We check for `total_count` being greater than 0 a few lines before, so I think this is fine.